### PR TITLE
Waku v2 Websocket support

### PR DESCRIPTION
**Pull request by rshiv**
This PR introduces web socket transport to wakunode and subsequent unit tests. 
Next iteration will introduce secure websocket capablities to the same.

Closes issue https://github.com/status-im/nim-waku/issues/756 and increment towards issue https://github.com/status-im/nim-waku/issues/434

The PR is ready to be merged

### CODE REVIEW

File: waku/v2/node/wakunode2.nim [hunk 1]
```diff
@@ -122,6 +124,9 @@ proc removeContentFilters(filters: var Filters, contentFilters: seq[ContentFilte
 template tcpEndPoint(address, port): auto =
   MultiAddress.init(address, tcpProtocol, port)
 
+template addWssFlag() =
+  MultiAddress.init(multiCodec("ws"))
```
Comment (jm-clius) [inline waku/v2/node/wakunode2.nim:128]: Shouldn't this be "wss" (codec 0x01DE) rather than "ws" (codec 0x01DD). You should also be able to initialise the MultiAddress just with the string "wss". It has an `init` function that will parse the `multiCodec` for you.

File: waku/v2/node/wakunode2.nim [hunk 2]
```diff
@@ -136,7 +141,10 @@ proc new*(T: type WakuNode, nodeKey: crypto.PrivateKey,
   ##
   let
     rng = crypto.newRng()
-    hostAddress = tcpEndPoint(bindIp, bindPort)
+    hostAddress = tcpEndPoint(bindIp, bindPort) 
+
+    hostAddressWithWss = hostAddress & addWssFlag.get()
```
Comment (jm-clius) [inline waku/v2/node/wakunode2.nim:146]: I don't think this is used anywhere? It should be possible to build a `switch` with both TCP and wss endpoints (use `withAddresses()` rather than `withAddress()`). Also, your host would need to use a different TCP port for the wss endpoint.

File: waku/v2/node/wakunode2.nim [hunk 3]
```diff
@@ -153,6 +161,17 @@ proc new*(T: type WakuNode, nodeKey: crypto.PrivateKey,
   for multiaddr in announcedAddresses:
     peerInfo.addrs.add(multiaddr) # Announced addresses in index > 0
   
+  # WSS switch builder code
+  let WssSwitch = SwitchBuilder
```
Comment (jm-clius) [inline waku/v2/node/wakunode2.nim:165]:
```suggestion
  let wssSwitch = SwitchBuilder
```

Variable names should be camelCase with first letter lowercase.

File: waku/v2/node/wakunode2.nim [hunk 4]
```diff
@@ -153,6 +161,17 @@ proc new*(T: type WakuNode, nodeKey: crypto.PrivateKey,
...
+    .new()
+    .withAddress(hostAddress)
+    .withRng(rng)
+    .withMplex()
+    .withTransport(proc (upgr: Upgrade): Transport = WsTransport.new(upgr))
+    .withNoise()
+    .build()
+
+
   var switch = newStandardSwitch(some(nodekey), hostAddress,
```
Comment (jm-clius) [inline waku/v2/node/wakunode2.nim:175]: I understand that it's easier to test with a dedicated switch for `wss`, but I don't think we want two switches (a single `switch` supports multiple transports and multiple host endpoints already). What's more your `wssSwitch` does not have a private key, etc. My suggestion: `wakunode2` should stop relying on `newStandardSwitch()` and instead use its own constructor that is based on that function. The only difference to what was before, is that wss endpoint and transport information _can_ also be included, if enabled in config. This constructor should preferably live in another module, like `waku_setup`.

File: waku/v2/node/wakunode2.nim (hunk 2 above)
Comment (D4nte) [reply waku/v2/node/wakunode2.nim:146]: I'd expect the host to be able to run both TCP and WSS at the same time (just in case it was not clear).

**Comment by rshiv** [comment]
> How is this going @rshiv? Would you mind updating PR description to explain current state and blocker? Feel free to include logs in here too with what is the current problem you are trying to figure out.

The integration is successful I will update the PR to reflect the changes for web socket support, and later add wss transport support as those would need tls certificate to work.

File: waku/v2/node/wakunode2.nim (hunk 2 above)
Comment (rshiv) [reply waku/v2/node/wakunode2.nim]: I was reading about the same, as per libp2p documentation this case of multiple transport is handled using circuit relay, however I am not sure, I am trying to get more information, and to check if there is some workaround.

File: waku/v2/node/wakunode2.nim [hunk 5]
```diff
@@ -138,28 +181,38 @@ proc new*(T: type WakuNode, nodeKey: crypto.PrivateKey,
   let
     rng = crypto.newRng()
     hostAddress = tcpEndPoint(bindIp, bindPort)
+    wsHostAddress = tcpEndPoint(bindIp, wsbindPort) & addWsFlag
     announcedAddresses = if extIp.isNone() or extPort.isNone(): @[]
```
Comment (rshiv) [inline waku/v2/node/wakunode2.nim:185]: The extIp is the tcp Ip, wakunode new function requires change for the new config to work.

File: waku/v2/node/config.nim [hunk 6]
```diff
@@ -47,6 +52,10 @@ type
       defaultValue: 50
       name: "max-connections" }: uint16
     
+    transport* {.
+      desc: "Specify transport to use for wakunode " &
```
Comment (oskarth) [inline waku/v2/node/config.nim:56]: This should allow for multiple transports. We don't want to add support for Websockets and remove support for TCP.

File: waku/v2/node/wakunode2.nim [hunk 7]
```diff
@@ -138,28 +181,38 @@ proc new*(T: type WakuNode, nodeKey: crypto.PrivateKey,
...
   for multiaddr in announcedAddresses:
     peerInfo.addrs.add(multiaddr) # Announced addresses in index > 0
   
-  var switch = newStandardSwitch(
-    some(nodekey),
-    hostAddress,
-    transportFlags = {ServerFlags.ReuseAddr},
-    rng = rng,
-    maxConnections = maxConnections)
+  var switch = newWakuSwitch(some(nodekey), wsHostAddress,
```
Comment (oskarth) [inline waku/v2/node/wakunode2.nim:207]: I think the existing formatting is fine and more clear, this change seems unrelated to PR

**Review by oskarth — COMMENTED** [review]
I see this PR is in draft mode. It'd be good to communicate clearly what is working and what is still a WIP. For example, with excerpt of logs showing how it works and/or how it doesn't.

Also seems to miss tests. Useful tests would be to ensure ws switch works in isolation, and that it works when a node supports tcp and ws (for example).

**Comment by oskarth** [comment]
>  later add wss transport support as those would need tls certificate to work.

We can do this in a separate PR once we basic capabilities done here

File: waku/v2/node/config.nim [hunk 8]
```diff
@@ -31,7 +31,12 @@ type
       desc: "TCP listening port."
       defaultValue: 60000
       name: "tcp-port" }: Port
-
+    
+    wsPort* {.
+      desc: "Web Socket listening port."
```
Comment (D4nte) [inline waku/v2/node/config.nim:36]:
```suggestion
      desc: "WebSocket listening port."
```

File: waku/v2/node/config.nim [hunk 9]
```diff
@@ -31,7 +31,12 @@ type
...
+      defaultValue: 1400
```
Comment (D4nte) [inline waku/v2/node/config.nim:37]: This is an odd value.

It's an http port. Commonly used value for http server run by a user (ie, port > 1024) are:
- 8080
- 8101
- 8000
- Something that looks like 80, the default http port.

File: waku/v2/node/wakunode2.nim [hunk 10]
```diff
@@ -122,14 +123,56 @@ proc removeContentFilters(filters: var Filters, contentFilters: seq[ContentFilte
...
+      .withAddress(address)
+      .withRng(rng)
+      .withMaxConnections(maxConnections)
+      .withMaxIn(maxIn)
+      .withMaxOut(maxOut)
+      .withMaxConnsPerPeer(maxConnsPerPeer)
+      .withMplex(inTimeout, outTimeout)
+      .withNoise()
+
+    if transport == "ws" or transport == "wss":
```
Comment (D4nte) [inline waku/v2/node/wakunode2.nim:160]: The addition of `wss` is not consistent across the PR. I'd suggest to leave it totally out for now and add it once you do a PR to add `wss`.

File: tests/v2/test_wakunode.nim [hunk 11]
```diff
@@ -478,7 +478,7 @@ procSuite "WakuNode":
       # invalid IP address
       discard parseRemotePeerInfo("/ip4/127.0.0.0.1/tcp/60002/p2p/16Uuu2HBmAcHvhLqQKwSSbX6BG5JLWUDRcaLVrehUVqpw7fz1hbYc")
     
-    expect ValueError:
+    expect LPError:
```
Comment (oskarth) [inline tests/v2/test_wakunode.nim:481]: Can ValueError no longer happen or was this wrong?

File: tests/v2/test_wakunode.nim [hunk 12]
```diff
@@ -1028,3 +1028,140 @@ procSuite "WakuNode":
...
+    await sleepAsync(2000.millis)
+
+
+    check:
+      (await completionFut.withTimeout(5.seconds)) == true
+    await node1.stop()
+    await node2.stop()
+
+
+asyncTest "Messages are relayed with one tcp and one multi transport (tcp and websocket)":
```
Comment (oskarth) [inline tests/v2/test_wakunode.nim:1078]:
```suggestion
asyncTest "Messages are relayed between nodes with multiple transports (TCP and Websockets)":
```

File: tests/v2/test_wakunode.nim [hunk 13]
```diff
@@ -1028,3 +1028,140 @@ procSuite "WakuNode":
...
+    await node2.publish(pubSubTopic, message)
+    await sleepAsync(2000.millis)
+
+
+    check:
+      (await completionFut.withTimeout(5.seconds)) == true
+    await node1.stop()
+    await node2.stop()
+
+asyncTest "Messages relay fails with one tcp and only websocket transport":
```
Comment (oskarth) [inline tests/v2/test_wakunode.nim:1122]:
```suggestion
asyncTest "Messages relaying fails with non-overlapping transports (TCP or Websockets)":
```

File: waku/v2/node/wakunode2.nim [hunk 14]
```diff
@@ -138,28 +144,43 @@ proc new*(T: type WakuNode, nodeKey: crypto.PrivateKey,
...
-    rng = rng,
-    maxConnections = maxConnections)
+  var switch = newWakuSwitch(some(nodekey),
+  hostAddress,
+  wsHostAddress, 
+  transportFlags = {ServerFlags.ReuseAddr},
+  rng = rng, 
+  maxConnections = maxConnections,
+  wsFlag = wsFlag)
+  # TODO Untested - verify behavior after switch interface change
```
Comment (oskarth) [inline waku/v2/node/wakunode2.nim:177]: What is this comment referring to?

Can write `wsFlag` as `wsEnabled` to make intent a bit more clear

**Review by oskarth — COMMENTED** [review]
Nice one! Definitely taking shaping :) Minor comments

**Comment by jm-clius** [comment]
Forgot to add: please update [CHANGELOG.md](https://github.com/status-im/nim-waku/blob/master/CHANGELOG.md) as well. Needed for release documentation.

File: tests/v2/test_wakunode.nim (hunk 11 above)
Comment (rshiv) [reply tests/v2/test_wakunode.nim:481]: Our implementation for check has changed and now we don't check the same and libp2p handles the scenario like other scenarios.

File: waku/v2/node/wakunode2.nim (hunk 14 above)
Comment (rshiv) [reply waku/v2/node/wakunode2.nim:177]: This is an old comment will remove the same.

File: waku/v2/utils/peers.nim [hunk 15]
```diff
@@ -44,22 +44,23 @@ proc init*(p: typedesc[RemotePeerInfo],
...
+## Check if wire Address is supported
+proc validWireAddr*(ma: MultiAddress): bool =
+  const
+    ValidTransports = mapOr(TCP, WebSockets)
+  if ValidTransports.match(ma) == true:
+    return true
   else:
-    raise newException(ValueError,
-                       "Invalid bootstrap node multi-address")
+    return false
```
Comment (jm-clius) [inline waku/v2/utils/peers.nim:54]: You can replace this whole conditional with:
```
return ValidTransports.match(ma)
```
This is simpler, since you are testing a boolean condition and returning the result of that condition.

File: waku/v2/utils/wakuswitch.nim [hunk 16]
```diff
@@ -0,0 +1,57 @@
+# Waku Switch utils.
+import
+  std/[options, tables, strutils, sequtils, os],
```
Comment (jm-clius) [inline waku/v2/utils/wakuswitch.nim:3]: Thanks for splitting this out! Are all of these imports necessary in the new module? Not sure, but I don't think e.g. `ping` and `enr` is used?
Imports can be messy, but good idea to try our best to not have unused/unnecessary imports.

**Review by jm-clius — COMMENTED** [review]
I've added a couple more comments, and would like to see the CHANGELOG updated too, but think we're v close to merging. 👍

File: waku/v2/utils/wakuswitch.nim [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+      ],
+    transportFlags: set[ServerFlags] = {},
+    rng = crypto.newRng(),
+    inTimeout: Duration = 5.minutes,
+    outTimeout: Duration = 5.minutes,
+    maxConnections = MaxConnections,
+    maxIn = -1,
+    maxOut = -1,
+    maxConnsPerPeer = MaxConnectionsPerPeer,
+    nameResolver: NameResolver = nil,
```
Comment (jm-clius) [inline waku/v2/utils/wakuswitch.nim:35]: This is not yet added in the `SwitchBuilder` below.

**Review by oskarth — APPROVED** [review]
LGTM

**Review by jm-clius — APPROVED** [review]
This looks good! Let's get it merged. Make sure you squash all commits into a single commit based on `master` (Github's default "Squash and Merge" should work).
